### PR TITLE
Feature/gal 993

### PR DESCRIPTION
**Pull request by Amar-NHS**
## Description

Ticket : https://nhsd-jira.digital.nhs.uk/browse/GAL-993

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    };
+    console.log(
+      "Preview of error acknowledgement response object attributes",
+      errorAckResponseObject
+    );
+    await sendMessageToQueue(errorAckResponseObject, QUEUE_URL, sqs, key);
+    await deleteObjectFromS3(bucket, key, s3);
+    console.log(`Successfully sent the message to SQS Queue: ${QUEUE_NAME}`);
+  } catch (error) {
+    console.error(`Lambda process was not successful in this instance`);
```
Comment (MichaelW-NHS) [inline terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js:45]: I think we're meant to be prefixing our error messages with Error:

File: terraform/main.tf [hunk 2]
```diff
@@ -1549,6 +1549,110 @@ module "send_single_notify_message_SQS_trigger" {
...
+module "fhir_validation_step4_error_bucket_lambda_trigger" {
+  source       = "./modules/lambda_trigger"
+  statement_id = "${var.environment}-AllowExecution-Step-4"
+  bucket_id    = module.inbound_nrds_galleritestresult_step4_error.bucket_id
+  bucket_arn   = module.inbound_nrds_galleritestresult_step4_error.bucket_arn
+  lambda_arn   = module.send_test_result_error_ack_queue_lambda.lambda_arn
+}
+
+# SNS Topic to publish and read test result ok Acknowledgement response
+module "test_result_ok_ack_topic" {
```
Comment (chenglawson1) [inline terraform/main.tf:1619]: This topic is not just for acknowledgement, it's also used for generating referral documents (see infrastructure diagram), better to rename so it is not ack specific.

File: terraform/main.tf [hunk 3]
```diff
@@ -1549,6 +1549,110 @@ module "send_single_notify_message_SQS_trigger" {
...
+  source      = "./modules/sns_topic"
+  environment = var.environment
+  name        = "testResultOkAckTopic"
```
Comment (chenglawson1) [inline terraform/main.tf:1622]: See above.

File: terraform/main.tf (hunk 2 above)
Comment (Amar-NHS) [reply terraform/main.tf:1619]: changed to "testResultTopic" as per diagram

File: terraform/main.tf (hunk 3 above)
Comment (Amar-NHS) [reply terraform/main.tf:1622]: changed to "testResultTopic" as per diagram

File: terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js (hunk 1 above)
Comment (Amar-NHS) [reply terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js:45]: Got two Error logs for the lambda handler, the second log below the one above has the prefix "Error:"

**Review by Abdul-NHS — APPROVED** [review]
LGTM

File: terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    const errorAckResponseObject = {
+      grail_fhir_result_id: grailFhirResultId,
+      ack_code: "fatal-error",
+    };
+    console.log(
+      "Preview of error acknowledgement response object attributes",
+      errorAckResponseObject
+    );
+    await sendMessageToQueue(errorAckResponseObject, QUEUE_URL, sqs, key);
+    await deleteObjectFromS3(bucket, key, s3);
```
Comment (chenglawson1) [inline terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js:42]: Is deleting the test result file from the bucket a requirement?

Comment (Amar-NHS) [reply terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js:42]: It's not a requirement, just followed what we have been doing with the previous lambdas that process the S3 event files. Would you like me to remove the deletion step?

Comment (chenglawson1) [reply terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js:42]: I'll check with Graham.

Comment (chenglawson1) [reply terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js:42]: Graham agrees we need to retain the failed test result file.

Comment (Amar-NHS) [reply terraform/src/sendTestResultErrorAckQueueLambda/sendTestResultErrorAckQueueLambda.js]: removed the deletion step from the sendTestResultErrorAckQueueLambda.js

**Review by chenglawson1 — APPROVED** [review]

**Review by avi312NHS — APPROVED** [review]

**Pull request merged**